### DISCUSSION TITLE
Generate flat road programmatically (only decal is fine)

### DISCUSSION THREAD

**Discussion by leonardo-panseri** [carla-simulator/carla#6858] [Q&A]
Hi, I'm trying to understand if CARLA offers a simple way to generate a flat road (~200m, with 2 lanes and some curves) programmatically to test behaviors of a regression model for a LKAS. I need to generate them programmatically because I want to explore the input space of the model and find where it starts to fail. I am currently using BeamNG that offers a feature for this (https://beamngpy.readthedocs.io/en/latest/beamngpy.html#beamngpy.Road), but I would like to switch to CARLA, as it is very well documented and more lightweigth.
I've tried CARLA OpenDrive Standalone Mode (https://carla.readthedocs.io/en/latest/adv_opendrive) but it doesn't generate road marks as of yet and this makes it impossible to test a camera-based LKAS.
Even generating only the mesh of the road without all the additional info for vehicle and pedestrian traffic would be okay.
Thank you for the help.

**Comment by Mayuemei** [comment]
![1](https://github.com/user-attachments/assets/c79ca4d0-a996-4b17-bbe8-985d4cd1281b)

May I ask if this issue has been resolved? I generated a road in OpenDrying format and successfully loaded it in Carla. However, when the vehicle is driving automatically on it, there is a problem and it will be destroyed by programs other than Python. The following error message appears. Here is the complete output of the program. Do you know the reason?


import carla
import time


def main():
    # Connect to Carla server
    client = carla.Client('localhost', 2000)
    client.set_timeout(10.0)

    # Load custom OpenDrive map
    with open(r'666.xodr', 'r') as od_file:
        data = od_file.read()
   
    try:
        # Generate the world with the custom OpenDrive map
        world = client.generate_opendrive_world(data)
        print("Custom OpenDrive map loaded successfully.")
    except RuntimeError as e:
        print("Error generating OpenDrive world:", e)
        return

    # Set weather to clear noon for blue sky
    weather = carla.WeatherParameters(
        cloudiness=0.0,
        precipitation=0.0,
        precipitation_deposits=0.0,
        wind_intensity=0.0,
        sun_azimuth_angle=0.0,
        sun_altitude_angle=90.0,
        fog_density=0.0,
        fog_distance=0.0,
        fog_falloff=0.0,
        wetness=0.0,
        scattering_intensity=0.0,
        mie_scattering_scale=0.0,
        rayleigh_scattering_scale=0.0331,
        dust_storm=0.0
    )
    world.set_weather(weather)

    # Clear the world of vehicles and pedestrians
    for actor in world.get_actors():
        if actor.type_id.startswith('vehicle.') or actor.type_id.startswith('walker.'):
            actor.destroy()

    # Get blueprint library
    blueprint_library = world.get_blueprint_library()
   
    # Select vehicle model
    vehicle_bp = blueprint_library.filter('model3')[0]

    # Set initial spawn point
    spawn_point = carla.Transform(carla.Location(x=3.0, y=0.0, z=1), carla.Rotation(pitch=0.0, yaw=0.0, roll=0.0))

    # Spawn vehicle
    vehicle = world.spawn_actor(vehicle_bp, spawn_point)
    print("Vehicle spawned.")

    # Enable autopilot
    vehicle.set_autopilot(True)
    print("Autopilot enabled.")

    # Set speed limit (if available)
    if hasattr(vehicle, 'set_speed_limit'):
        vehicle.set_speed_limit(10)  # Set a speed limit of 30 km/h

    snap = world.wait_for_tick()
    init_frame = snap.frame
    run_frame = 0
    # Get the map object
    map = world.get_map()
    print(f"spawn ego vehicle at: {spawn_point}")

    while run_frame < 100000:
        try:
            snap = world.wait_for_tick()
            run_frame = snap.frame - init_frame

            # Set spectator to follow the ego vehicle
            spectator = world.get_spectator()
            ego_tf = vehicle.get_transform()
            spectator_tf = carla.Transform(ego_tf.location, ego_tf.rotation)
            spectator_tf.location += ego_tf.get_forward_vector() * (-10)
            spectator_tf.location += ego_tf.get_up_vector() * 3
            spectator.set_transform(spectator_tf)

            # Get the waypoint at the vehicle's current location
            transform = vehicle.get_transform()
            waypoint = map.get_waypoint(transform.location)
            print(f"-- Current waypoint = {waypoint}")

            # Get current position's lane information
            road_id = waypoint.road_id
            section_id = waypoint.section_id
            lane_id = waypoint.lane_id
            lane_type = waypoint.lane_type
            s = waypoint.s
            print(
                f" -road_id = {road_id}, section_id = {section_id}, lane_id = {lane_id}, lane_type = {lane_type}, s = {s}")

            # Get nearby waypoints
            left_wp = waypoint.get_left_lane()  # Get waypoints on the left lane
            right_wp = waypoint.get_right_lane()  # Get waypoints on the right lane
            print(f" -left_wp = {left_wp}")
            print(f" -right_wp = {right_wp}")

        except RuntimeError as e:
            print("Error:", e)
            break

     # Destroy vehicle
    is_destroyed = vehicle.destroy()
    if is_destroyed:
        print("Ego vehicle has been destroyed successfully")



if __name__ == '__main__':
    main()




Custom OpenDrive map loaded successfully.
Vehicle spawned.
WARNING: cannot parse georeference: ''. Using default values. 
WARNING: No InMemoryMap cache found. Setting up local map. This may take a while... 
Autopilot enabled.
spawn ego vehicle at: Transform(Location(x=3.000000, y=0.000000, z=1.000000), Rotation(pitch=0.000000, yaw=0.000000, roll=0.000000))
-- Current waypoint = Waypoint(Transform(Location(x=3.001390, y=1.750000, z=0.000000), Rotation(pitch=0.000000, yaw=-0.000000, roll=0.000000)))
 -road_id = 1, section_id = 0, lane_id = -1, lane_type = Driving, s = 3.001390218734761
 -left_wp = Waypoint(Transform(Location(x=3.001390, y=-1.750000, z=0.000000), Rotation(pitch=360.000000, yaw=180.000000, roll=0.000000)))
 -right_wp = Waypoint(Transform(Location(x=3.001390, y=4.000000, z=0.000000), Rotation(pitch=0.000000, yaw=-0.000000, roll=0.000000)))
WARNING: cannot parse georeference: ''. Using default values. 
-- Current waypoint = Waypoint(Transform(Location(x=3.001390, y=1.750000, z=0.000000), Rotation(pitch=0.000000, yaw=-0.000000, roll=0.000000)))
 -road_id = 1, section_id = 0, lane_id = -1, lane_type = Driving, s = 3.001390218734761
 -left_wp = Waypoint(Transform(Location(x=3.001390, y=-1.750000, z=0.000000), Rotation(pitch=360.000000, yaw=180.000000, roll=0.000000)))
 -right_wp = Waypoint(Transform(Location(x=3.001390, y=4.000000, z=0.000000), Rotation(pitch=0.000000, yaw=-0.000000, roll=0.000000)))
Error: trying to operate on a destroyed actor; an actor's function was called, but the actor is already destroyed.
Traceback (most recent call last):
  File "D:/test/test2.py", line 135, in <module>
    main()
  File "D:/test/test2.py", line 128, in main
    is_destroyed = vehicle.destroy()
RuntimeError: trying to operate on a destroyed actor; an actor's function was called, but the actor is already destroyed.
